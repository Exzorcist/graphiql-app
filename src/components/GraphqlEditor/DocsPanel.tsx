import { ComponentProps } from 'react';
import { cn } from '@/utils';

function DocsPanel({ className, ...delegated }: ComponentProps<'div'>) {
  return (
    <div
      className={cn('bg-editor-primary flex items-center justify-center h-full w-full', className)}
      {...delegated}
    >
      Docs goes here
    </div>
  );
}

export default DocsPanel;
