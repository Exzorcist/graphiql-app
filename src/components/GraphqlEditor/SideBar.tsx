import Prettifying from '@/components/GraphqlEditor/Prettifying/Prettifying';

function SideBar() {
  return (
    <aside className="w-[50px] h-full bg-editor-primary border-editor-border border-r py-4 px-2">
      <Prettifying />
    </aside>
  );
}

export default SideBar;
