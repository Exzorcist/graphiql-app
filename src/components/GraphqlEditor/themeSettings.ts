const commmonEditorThemeSettings = {
  fontFamily: 'var(--font-mono)',
  gutterForeground: 'rgb(var(--editor-line-numbers-color))',
  selection: 'rgb(var(--editor-code-selection-color))',
};

export const primaryEditorThemeSettings = {
  ...commmonEditorThemeSettings,
  background: 'rgb(var(--editor-primary-color))',
  gutterBackground: 'rgb(var(--editor-gutter-color))',
};

export const secondaryEditorThemeSettings = {
  ...commmonEditorThemeSettings,
  background: 'rgb(var(--editor-secondary-color))',
  gutterBackground: 'rgb(var(--editor-secondary-color))',
};
