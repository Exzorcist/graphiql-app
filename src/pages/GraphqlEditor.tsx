function GraphqlEditor() {
  return <div>This is a GraphqlEditor page</div>;
}

export default GraphqlEditor;
