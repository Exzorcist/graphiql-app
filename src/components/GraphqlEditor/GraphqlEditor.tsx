import { Panel, PanelGroup } from 'react-resizable-panels';
import { useCallback, useRef, useState } from 'react';
import { Transition } from 'react-transition-group';
import PanelResizeHandle from '../PanelResizeHandle/PanelResizeHandle';
import GraphqlToolsPanel from './GraphqlToolsPanel/GraphqlToolsPanel';
import SchemaDocsDrawer from './SchemaDocs/SchemaDocsDrawer';
import SchemaDocsPanel from './SchemaDocs/SchemaDocsPanel';
import EndpointField from './EndpointField/EndpointField';
import ResponsePanel from './ResponsePanel/ResponsePanel';
import RequestPanel from './RequestPanel/RequestPanel';
import { useBreakpoint } from '@/hooks/useBreakpoint';
import SideBar from './SideBar';
import { cn } from '@/utils';

const PANEL_GROUP_ID = 'graphql-editor-panel-group';
const QUERY_EDITOR_PANEL_MIN_SIZE = 20;
const RESPONSE_PANEL_MIN_SIZE = 20;

function GraphqlEditor() {
  const [showDocs, setShowDocs] = useState(false);
  const [showTools, setShowTools] = useState(false);
  const nodeRef = useRef<HTMLDivElement | null>(null);
  const isLaptop = useBreakpoint('min-laptop');

  const handleDocsClick = useCallback(() => {
    setShowDocs((prev) => !prev);
  }, []);

  const setShowDocsMemo = useCallback((state: boolean) => setShowDocs(state), []);
  const setShowToolsMemo = useCallback((state: boolean) => setShowTools(state), []);

  return (
    <div className="flex h-full w-full text-editor-text-color bg-editor-primary font-editor-font-family">
      {isLaptop && <SideBar />}
      {!isLaptop && <SchemaDocsDrawer open={showDocs} onOpenChange={setShowDocsMemo} />}
      <div className="h-full flex w-full flex-col">
        <div className="bg-editor-primary px-4 py-5 border-editor-border border-b">
          <EndpointField onSchemaClick={handleDocsClick} isSchemaOpen={showDocs} />
        </div>
        <Transition in={isLaptop ? showDocs : showTools} timeout={150} nodeRef={nodeRef}>
          {(state) => {
            const panelClassName = cn(
              (state === 'entering' || state === 'exiting') && 'animated-panel'
            );

            const currentPanel = isLaptop ? (
              <SchemaDocsPanel
                id="docsPanel"
                order={3}
                show={showDocs}
                onShowChange={setShowDocsMemo}
                panelClassName={panelClassName}
              />
            ) : (
              <GraphqlToolsPanel
                id="toolsPanel"
                order={3}
                onShowChange={setShowToolsMemo}
                panelGroupId={PANEL_GROUP_ID}
                panelClassName={panelClassName}
              />
            );

            return (
              <div className="w-full h-full" ref={nodeRef}>
                <PanelGroup
                  id={PANEL_GROUP_ID}
                  autoSaveId={PANEL_GROUP_ID}
                  direction={isLaptop ? 'horizontal' : 'vertical'}
                >
                  <Panel
                    id="requestPanel"
                    order={1}
                    minSize={QUERY_EDITOR_PANEL_MIN_SIZE}
                    className={panelClassName}
                  >
                    <RequestPanel />
                  </Panel>
                  <PanelResizeHandle />
                  <Panel
                    id="responsePanel"
                    order={2}
                    minSize={RESPONSE_PANEL_MIN_SIZE}
                    className={panelClassName}
                  >
                    <ResponsePanel />
                  </Panel>
                  <PanelResizeHandle />
                  {currentPanel}
                </PanelGroup>
              </div>
            );
          }}
        </Transition>
      </div>
    </div>
  );
}

export default GraphqlEditor;
