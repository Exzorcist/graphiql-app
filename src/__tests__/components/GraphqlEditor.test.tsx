import GraphqlEditor from '@/components/GraphqlEditor/GraphqlEditor';
import { customRender as render } from '../test-utils';

test('GraphqlEditor', () => {
  render(<GraphqlEditor />);

  expect(document.body).toMatchSnapshot();
});
