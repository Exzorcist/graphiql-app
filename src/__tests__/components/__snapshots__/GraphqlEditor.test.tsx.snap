// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GraphqlEditor 1`] = `
<body>
  <div>
    <div
      class="flex h-full w-full max-w-[100vw] text-editor-text-color bg-editor-primary font-sans selection:bg-editor-code-selection"
    >
      <div
        class="h-full flex w-full flex-col min-w-0"
      >
        <div
          class="bg-editor-primary px-4 pt-5  sm:flex sm:items-center sm:gap-5 lg:block"
        >
          <form
            class="h-10 flex w-full bg-editor-secondary border-editor-border border rounded [&:has(input:focus)]:border-editor-accent"
          >
            <div
              class="w-full h-full flex "
            >
              <input
                class="w-full h-full bg-transparent outline-none pl-4"
                placeholder="Enter URL"
                value=""
              />
              <div
                class="bottom-0 right-4 h-full flex items-center pr-4 pl-4"
              >
                <button
                  class="py-[5px] px-2 transition rounded min-w-fit cursor-pointer hover:bg-editor-accent/[.15] hover:text-editor-accent"
                  role="button"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="hidden lg:inline"
                  >
                    Schema
                  </span>
                  <svg
                    aria-hidden="true"
                    class="h-6 w-6 lg:hidden"
                    data-slot="icon"
                    fill="none"
                    stroke="currentColor"
                    stroke-width="1.5"
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M12 6.042A8.967 8.967 0 0 0 6 3.75c-1.052 0-2.062.18-3 .512v14.25A8.987 8.987 0 0 1 6 18c2.305 0 4.408.867 6 2.292m0-14.25a8.966 8.966 0 0 1 6-2.292c1.052 0 2.062.18 3 .512v14.25A8.987 8.987 0 0 0 18 18a8.967 8.967 0 0 0-6 2.292m0-14.25v14.25"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                    />
                  </svg>
                </button>
              </div>
            </div>
            <button
              class="py-[5px] transition min-w-fit cursor-pointer bg-editor-accent hover:bg-editor-accent-light text-white px-5 rounded-none rounded-r"
              role="button"
              tabindex="0"
              type="submit"
            >
              <svg
                aria-hidden="true"
                class="w-6 h-6"
                data-slot="icon"
                fill="none"
                stroke="currentColor"
                stroke-width="1.5"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5.25 5.653c0-.856.917-1.398 1.667-.986l11.54 6.347a1.125 1.125 0 0 1 0 1.972l-11.54 6.347a1.125 1.125 0 0 1-1.667-.986V5.653Z"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </button>
          </form>
        </div>
        <div
          class="w-full h-full"
        >
          <div
            class="relative"
            data-panel-group=""
            data-panel-group-direction="vertical"
            data-panel-group-id="graphql-editor-panel-group"
            style="display: flex; flex-direction: column; height: 100%; overflow: hidden; width: 100%;"
          >
            <div
              class=""
              data-panel=""
              data-panel-group-id="graphql-editor-panel-group"
              data-panel-id="requestPanel"
              data-panel-size="1.0"
              style="flex-basis: 0px; flex-grow: 1; flex-shrink: 1; overflow: hidden;"
            >
              <div
                class="w-full h-full"
              >
                <div
                  class="relative"
                  data-panel-group=""
                  data-panel-group-direction="vertical"
                  data-panel-group-id="request-panel-group"
                  style="display: flex; flex-direction: column; height: 100%; overflow: hidden; width: 100%;"
                >
                  <div
                    class=""
                    data-panel=""
                    data-panel-group-id="request-panel-group"
                    data-panel-id="queryPanel"
                    data-panel-size="1.0"
                    style="flex-basis: 0px; flex-grow: 1; flex-shrink: 1; overflow: hidden;"
                  >
                    <div
                      class="h-full relative"
                    >
                      <div
                        class="w-full absolute top-0 left-0 px-4 flex items-center border-b-editor-border border-b"
                        style="height: 50px;"
                      >
                        <div
                          class="flex w-full"
                        >
                          <button
                            class="py-[5px] px-2 transition rounded min-w-fit cursor-pointer hover:bg-editor-secondary hover:text-editor-accent text-editor-accent bg-editor-secondary"
                            role="button"
                            tabindex="0"
                            type="button"
                          >
                            Request
                          </button>
                          <button
                            class="py-[5px] px-2 transition rounded min-w-fit cursor-pointer hover:bg-editor-secondary hover:text-editor-accent ml-auto"
                            role="button"
                            tabindex="0"
                            title="Prettify"
                            type="button"
                          >
                            <div
                              class="flex items-center gap-2"
                            >
                              <span
                                class="hidden sm:inline"
                              >
                                Prettify
                              </span>
                              <svg
                                aria-hidden="true"
                                class="w-7 h-7"
                                data-slot="icon"
                                fill="none"
                                stroke="currentColor"
                                stroke-width="1.5"
                                viewBox="0 0 24 24"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <path
                                  d="M14.25 6.087c0-.355.186-.676.401-.959.221-.29.349-.634.349-1.003 0-1.036-1.007-1.875-2.25-1.875s-2.25.84-2.25 1.875c0 .369.128.713.349 1.003.215.283.401.604.401.959v0a.64.64 0 0 1-.657.643 48.39 48.39 0 0 1-4.163-.3c.186 1.613.293 3.25.315 4.907a.656.656 0 0 1-.658.663v0c-.355 0-.676-.186-.959-.401a1.647 1.647 0 0 0-1.003-.349c-1.036 0-1.875 1.007-1.875 2.25s.84 2.25 1.875 2.25c.369 0 .713-.128 1.003-.349.283-.215.604-.401.959-.401v0c.31 0 .555.26.532.57a48.039 48.039 0 0 1-.642 5.056c1.518.19 3.058.309 4.616.354a.64.64 0 0 0 .657-.643v0c0-.355-.186-.676-.401-.959a1.647 1.647 0 0 1-.349-1.003c0-1.035 1.008-1.875 2.25-1.875 1.243 0 2.25.84 2.25 1.875 0 .369-.128.713-.349 1.003-.215.283-.4.604-.4.959v0c0 .333.277.599.61.58a48.1 48.1 0 0 0 5.427-.63 48.05 48.05 0 0 0 .582-4.717.532.532 0 0 0-.533-.57v0c-.355 0-.676.186-.959.401-.29.221-.634.349-1.003.349-1.035 0-1.875-1.007-1.875-2.25s.84-2.25 1.875-2.25c.37 0 .713.128 1.003.349.283.215.604.401.96.401v0a.656.656 0 0 0 .658-.663 48.422 48.422 0 0 0-.37-5.36c-1.886.342-3.81.574-5.766.689a.578.578 0 0 1-.61-.58v0Z"
                                  stroke-linecap="round"
                                  stroke-linejoin="round"
                                />
                              </svg>
                            </div>
                          </button>
                        </div>
                      </div>
                      <div
                        class="cm-theme h-full flex flex-col"
                        data-scrollbar-gutter="true"
                        data-testid="editor-area"
                        style="padding-top: 50px;"
                      >
                        <div
                          class="cm-editor ͼ1 ͼ3 ͼ4 ͼ5p ͼ4s ͼ4q ͼ16"
                        >
                          <div
                            aria-live="polite"
                            class="cm-announced"
                          />
                          <div
                            class="cm-scroller"
                            tabindex="-1"
                          >
                            <div
                              aria-hidden="true"
                              class="cm-gutters"
                              style="min-height: 14px; position: sticky;"
                            >
                              <div
                                class="cm-gutter cm-lineNumbers"
                              >
                                <div
                                  class="cm-gutterElement"
                                  style="height: 0px; visibility: hidden; pointer-events: none;"
                                >
                                  9
                                </div>
                                <div
                                  class="cm-gutterElement"
                                  style="height: 14px;"
                                >
                                  1
                                </div>
                              </div>
                            </div>
                            <div
                              aria-autocomplete="list"
                              aria-multiline="true"
                              autocapitalize="off"
                              autocorrect="off"
                              class="cm-content"
                              contenteditable="true"
                              role="textbox"
                              spellcheck="false"
                              style="tab-size: 4;"
                              translate="no"
                            >
                              <div
                                class="cm-line"
                              >
                                <br />
                              </div>
                            </div>
                            <div
                              aria-hidden="true"
                              class="cm-layer cm-layer-above cm-cursorLayer"
                              style="z-index: 150; animation-duration: 1200ms;"
                            />
                            <div
                              aria-hidden="true"
                              class="cm-layer cm-selectionLayer"
                              style="z-index: -2;"
                            />
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="z-50 group"
              data-panel-group-direction="vertical"
              data-panel-group-id="graphql-editor-panel-group"
              data-panel-resize-handle-enabled="true"
              data-panel-resize-handle-id=":r7:"
              data-resize-handle=""
              role="separator"
              style="cursor: ns-resize; user-select: none;"
              tabindex="0"
            >
              <div
                class="_ResizeHandleGutter_dd249a"
              >
                <div
                  class="_ResizeHandleDragArea_dd249a peer"
                />
                <div
                  class="_ResizeHandleHoverArea_dd249a peer-hover:delay-150 peer-hover:opacity-100 group-data-[resize-handle-active]:opacity-100"
                />
              </div>
            </div>
            <div
              class=""
              data-panel=""
              data-panel-group-id="graphql-editor-panel-group"
              data-panel-id="responsePanel"
              data-panel-size="1.0"
              style="flex-basis: 0px; flex-grow: 1; flex-shrink: 1; overflow: hidden;"
            >
              <div
                class="h-full relative"
              >
                <div
                  class="w-full absolute top-0 left-0 px-4 flex items-center border-b-editor-border border-b"
                  style="height: 50px;"
                >
                  <div
                    class="flex h-full items-center w-full"
                  >
                    <button
                      class="py-[5px] px-2 transition rounded min-w-fit cursor-pointer hover:bg-editor-secondary hover:text-editor-accent text-editor-accent bg-editor-secondary mr-5"
                      role="button"
                      tabindex="0"
                      type="button"
                    >
                      Response
                    </button>
                    <div
                      class="ml-auto flex h-full py-3 gap-3 items-center"
                    />
                  </div>
                </div>
                <div
                  class="cm-theme h-full flex flex-col"
                  data-testid="editor-area"
                  style="padding-top: 50px;"
                >
                  <div
                    class="cm-editor ͼ1 ͼ3 ͼ4 ͼ5q ͼ52 ͼ4q ͼ16"
                  >
                    <div
                      aria-live="polite"
                      class="cm-announced"
                    />
                    <div
                      class="cm-scroller"
                      tabindex="-1"
                    >
                      <div
                        aria-hidden="true"
                        class="cm-gutters"
                        style="min-height: 14px; position: sticky;"
                      >
                        <div
                          class="cm-gutter cm-lineNumbers"
                        >
                          <div
                            class="cm-gutterElement"
                            style="height: 0px; visibility: hidden; pointer-events: none;"
                          >
                            9
                          </div>
                          <div
                            class="cm-gutterElement"
                            style="height: 14px;"
                          >
                            1
                          </div>
                        </div>
                      </div>
                      <div
                        aria-autocomplete="list"
                        aria-multiline="true"
                        aria-readonly="true"
                        autocapitalize="off"
                        autocorrect="off"
                        class="cm-content"
                        contenteditable="true"
                        role="textbox"
                        spellcheck="false"
                        style="tab-size: 4;"
                        translate="no"
                      >
                        <div
                          class="cm-line"
                        >
                          <br />
                        </div>
                      </div>
                      <div
                        aria-hidden="true"
                        class="cm-layer cm-layer-above cm-cursorLayer"
                        style="z-index: 150; animation-duration: 1200ms;"
                      />
                      <div
                        aria-hidden="true"
                        class="cm-layer cm-selectionLayer"
                        style="z-index: -2;"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="z-50 group"
              data-panel-group-direction="vertical"
              data-panel-group-id="graphql-editor-panel-group"
              data-panel-resize-handle-enabled="true"
              data-panel-resize-handle-id=":r9:"
              data-resize-handle=""
              role="separator"
              style="cursor: ns-resize; user-select: none;"
              tabindex="0"
            >
              <div
                class="_ResizeHandleGutter_dd249a"
              >
                <div
                  class="_ResizeHandleDragArea_dd249a peer"
                />
                <div
                  class="_ResizeHandleHoverArea_dd249a peer-hover:delay-150 peer-hover:opacity-100 group-data-[resize-handle-active]:opacity-100"
                />
              </div>
            </div>
            <div
              class=""
              data-panel=""
              data-panel-collapsible="true"
              data-panel-group-id="graphql-editor-panel-group"
              data-panel-id="toolsPanel"
              data-panel-size="0.0"
              style="flex-basis: 0px; flex-grow: 0; flex-shrink: 1; overflow: hidden;"
            >
              <div
                class="h-full relative"
              >
                <div
                  class="gap-6 w-full absolute top-0 left-0 px-4 flex items-center h-[50px]"
                >
                  <button
                    class="py-[5px] px-2 transition rounded min-w-fit cursor-pointer hover:bg-editor-secondary hover:text-editor-accent"
                    role="button"
                    tabindex="0"
                    type="button"
                  >
                    Variables
                  </button>
                  <button
                    class="py-[5px] px-2 transition rounded min-w-fit cursor-pointer hover:bg-editor-secondary hover:text-editor-accent"
                    role="button"
                    tabindex="0"
                    type="button"
                  >
                    Headers
                  </button>
                  <button
                    class="py-[5px] px-2 transition rounded min-w-fit cursor-pointer hover:bg-editor-secondary hover:text-editor-accent ml-auto"
                    role="button"
                    tabindex="0"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="h-6 w-6"
                      data-slot="icon"
                      fill="currentColor"
                      viewBox="0 0 24 24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        clip-rule="evenodd"
                        d="M11.47 7.72a.75.75 0 0 1 1.06 0l7.5 7.5a.75.75 0 1 1-1.06 1.06L12 9.31l-6.97 6.97a.75.75 0 0 1-1.06-1.06l7.5-7.5Z"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </button>
                </div>
                <div
                  class="h-full pt-[50px]"
                >
                  <div
                    class="cm-theme h-full flex flex-col"
                    data-scrollbar-gutter="true"
                    data-testid="editor-area"
                  >
                    <div
                      class="cm-editor ͼ1 ͼ3 ͼ4 ͼ5r ͼ5c ͼ4q ͼ16"
                    >
                      <div
                        aria-live="polite"
                        class="cm-announced"
                      />
                      <div
                        class="cm-scroller"
                        tabindex="-1"
                      >
                        <div
                          aria-hidden="true"
                          class="cm-gutters"
                          style="min-height: 42px; position: sticky;"
                        >
                          <div
                            class="cm-gutter cm-lineNumbers"
                          >
                            <div
                              class="cm-gutterElement"
                              style="height: 0px; visibility: hidden; pointer-events: none;"
                            >
                              9
                            </div>
                            <div
                              class="cm-gutterElement"
                              style="height: 14px;"
                            >
                              1
                            </div>
                            <div
                              class="cm-gutterElement"
                              style="height: 14px;"
                            >
                              2
                            </div>
                            <div
                              class="cm-gutterElement"
                              style="height: 14px;"
                            >
                              3
                            </div>
                          </div>
                        </div>
                        <div
                          aria-autocomplete="list"
                          aria-multiline="true"
                          autocapitalize="off"
                          autocorrect="off"
                          class="cm-content"
                          contenteditable="true"
                          role="textbox"
                          spellcheck="false"
                          style="tab-size: 4;"
                          translate="no"
                        >
                          <div
                            class="cm-line"
                          >
                            {
                          </div>
                          <div
                            class="cm-line"
                          >
                            <br />
                          </div>
                          <div
                            class="cm-line"
                          >
                            }
                          </div>
                        </div>
                        <div
                          aria-hidden="true"
                          class="cm-layer cm-layer-above cm-cursorLayer"
                          style="z-index: 150; animation-duration: 1200ms;"
                        />
                        <div
                          aria-hidden="true"
                          class="cm-layer cm-selectionLayer"
                          style="z-index: -2;"
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
